packages/remix-dev/config/flat-routes.ts

```diff
@@ -0,0 +1,291 @@
+ * @param {string[]} routePaths - The absolute route paths.
+ * @param {string} [prefix=routes] - The prefix to strip off of the routes.
+ */
+export function flatRoutesUniversal(
+  appDirectory: string,
+  routePaths: string[],
+  prefix: string = "routes"
+): RouteManifest {
+  let routeMap = getRouteMap(appDirectory, routePaths, prefix);
+
+  let uniqueRoutes = new Map<string, string>();
+
+  function defineNestedRoutes(
+    defineRoute: DefineRouteFunction,
+    parentId?: string
+  ): void {
+    let childRoutes = Array.from(routeMap.values()).filter(
+      (routeInfo) => routeInfo.parentId === parentId
+    );
+    let parentRoute = parentId ? routeMap.get(parentId) : undefined;
+    let parentRoutePath = parentRoute?.path ?? "/";
+    for (let childRoute of childRoutes) {
+      let routePath = childRoute.path?.slice(parentRoutePath.length) ?? "";
+      // remove leading slash
+      routePath = routePath.replace(/^\//, "");
+      let index = childRoute.index;
+      let fullPath = childRoute.path;
+
+      let uniqueRouteId = (fullPath || "") + (index ? "?index" : "");
+      if (uniqueRouteId) {
+        let conflict = uniqueRoutes.get(uniqueRouteId);
+        if (conflict) {
+          throw new Error(
+            `Path ${JSON.stringify(fullPath)} defined by route ${JSON.stringify(
+              childRoute.id
+            )} conflicts with route ${JSON.stringify(conflict)}`
+          );
+        }
+        uniqueRoutes.set(uniqueRouteId, childRoute.id);
+      }
+      if (index) {
+        let invalidChildRoutes = Object.values(routeMap).filter(
+          (routeInfo) => routeInfo.parentId === childRoute.id
+        );
+
+        if (invalidChildRoutes.length > 0) {
+          throw new Error(
+            `Child routes are not allowed in index routes. Please remove child routes of ${childRoute.id}`
+          );
+        }
+
+        defineRoute(routePath, routeMap.get(childRoute.id!)!.file, {
+          index: true,
+        });
+      } else {
+        defineRoute(routePath, routeMap.get(childRoute.id!)!.file, () => {
+          defineNestedRoutes(defineRoute, childRoute.id);
+        });
+      }
+    }
+  }
+
+  let routes = defineRoutes(defineNestedRoutes);
+
+  return routes;
+}
+
+function isIndexRoute(routeId: string) {
+  return routeId.endsWith("_index");
+}
+
+type SubState = "NORMAL" | "ESCAPE" | "OPTIONAL" | "OPTIONAL_ESCAPE";
+
+export function getRouteSegments(routeId: string) {
+  let routeSegments: string[] = [];
+  let index = 0;
+  let routeSegment = "";
+  let state: "START" | "PATH" = "START";
+  let subState: SubState = "NORMAL";
+  let pushRouteSegment = (routeSegment: string) => {
+    if (routeSegment) {
+      if (routeSegment.includes("/")) {
+        throw new Error(
+          `Route segment cannot contain a slash: ${routeSegment} (in route ${routeId})`
+        );
+      }
+      routeSegments.push(routeSegment);
+    }
```

Comment:
nit: you can unnest most of this with a precondition / early return:
```suggestion
    if (!routeSegment) return
    if (routeSegment.includes("/")) {
      throw new Error(
        `Route segment cannot contain a slash: ${routeSegment} (in route ${routeId})`
      );
    }
    routeSegments.push(routeSegment);
```